### ISSUE TITLE
Calling `from_geocode` occasionally produces "list index out of range" error

### ISSUE THREAD

**Issue by scottishbee** [gmplot/gmplot#58]
I've been using gmplot to draw running routes (via tcx files).  Until recently I could loop through a file list and generate the new html plots.

But now the loop will seemingly-spontaneously fail and generate the following error:

> 
>      40             'http://maps.googleapis.com/maps/api/geocode/json?address="%s"' % location_string)
>      41         geocode = json.loads(geocode.text)
> ---> 42         latlng_dict = geocode['results'][0]['geometry']['location']
>      43         return latlng_dict['lat'], latlng_dict['lng']
>      44 
> 
> IndexError: list index out of range

I was able to isolate it to the call for 
`gmap = gmplot.GoogleMapPlotter.from_geocode("San Francisco")`

For example, running the following code will result in throwing an error at some point (usually in the 8-14 range):

```
for i in range(25):
    print(i)
    gmap = gmplot.GoogleMapPlotter.from_geocode("San Francisco")

```

I can wrap it in a try/except but that merely delays the break.

My naive guess is that I'm breaking some API limit.

**Comment by mriffey** [comment]
I see the same error and I just started using the API *today*. Ive made maybe 20-30 calls to it so far.

**Comment by vgm64** [comment]
You may have more success supplying your own API key. The limits are
assumable higher(?).

**Comment by mriffey** [comment]
I am using my own key - which is new today:)

I figured out what appears to be the cause. See the 2 sets of ###### below. 

Mark



```
import googlemaps, gmplot, webbrowser, os, json

gmaps = googlemaps.Client(key='your api key') 
geocode_result = gmaps.geocode('100 Broadway Street, Missoula, MT')

#######
gmap = gmplot.GoogleMapPlotter.from_geocode("Missoula")    # IF THIS LINE IS HERE, IT THROWS THE ERROR
#######

geom = geocode_result[0]['geometry']
loc = geom['location']
lat = loc['lat']
lng = loc['lng']

#######
gmap = gmplot.GoogleMapPlotter.from_geocode("Missoula")    # IF THIS LINE IS HERE, IT WORKS
#######

hidden_gem_lat, hidden_gem_lon = lat,lng
gmap.marker(hidden_gem_lat, hidden_gem_lon, 'cornflowerblue')

# Draw
gmap.draw("my_map.html")

filename = 'file:///'+os.getcwd()+'/' + 'my_map.html'
webbrowser.open_new_tab(filename)
```

**Comment by vgm64** [comment]
I see that moving a line can cause the error to not be thrown, but that's
not explaining the problem. It kinda seems like having two geocoding
requests back to back is to blame, which potentially seems reasonable? I'm
not sure what kind of interaction the packages you're using might have. Any
thoughts on a general fix for those experiencing this?

**Comment by mriffey** [comment]
I agree that it appears that back to back requests are the cause and that perhaps indicates there is some sort of initialization/clean up isn't happening which, if done, would allow a series of 2 or more geocoding calls to work properly. I think it's reasonable to expect such a series of calls to work. I am not sure that the module or my code is supposed to be done that clean up, nor do I know what that cleanup might be - if that's the issue. 

BTW, the code I posted is the complete code for this project at the moment.

**Comment by Tsingis** [comment]
Any solutions to this?

**Comment by vgm64** [comment]
I think we still stand at: multiple packages being used together causes something to break due to back-to-back geocoding calls.

If you have a new example that reproduces the issue, please share. Isolating the issue to this repo is a good next step.

**Comment by Tsingis** [comment]
`import json
import urllib.request as request
from numpy import cumsum
import gmplot

busLine = '25'
url = 'http://data.itsfactory.fi/journeys/api/1/routes'
req = request.urlopen(url)
if req.getcode() == 200:
    data = json.loads(req.read().decode())
body = data['body']
coords = []
for n in range(0, len(body)):
    if body[n]['lineUrl'].split('/')[-1] == busLine:
        coords.append(body[n]['geographicCoordinateProjection'])
co = coords[0].split(':') # Split coord values
# Split coords into lats and lons
lats = []
lons = []
for n in range(0, len(co)):
    lats.append(int(co[n].split(',')[0]))
    lons.append(int(co[n].split(',')[1]))
# Compute lats and lons from differences
lats = [lats[0]]+[i*(-1) for i in lats[1:]]
lats = list(cumsum(lats))
lats = [round(1e-5*i, 5) for i in lats]
lons = [lons[0]]+[i*(-1) for i in lons[1:]]
lons = list(cumsum(lons))
lons = [round(1e-5*i, 5) for i in lons]

gmap = gmplot.GoogleMapPlotter.from_geocode('Tampere', 15)
gmap.plot(lats, lons, 'steelblue', edge_width=3)
gmap.draw('map.html')
`
Pretty long reproducible example but with this one I get errors pretty often.

**Comment by vgm64** [comment]
So you would run this in, say, a bash for loop? Can you post your traceback for completeness?

**Comment by sandrotosi** [comment]
the issue is pretty obvious looking at the code:

```
    @classmethod
    def geocode(self, location_string):
        geocode = requests.get(
            'http://maps.googleapis.com/maps/api/geocode/json?address="%s"' % location_string)
        geocode = json.loads(geocode.text)
        latlng_dict = geocode['results'][0]['geometry']['location']
        return latlng_dict['lat'], latlng_dict['lng']
```

first you need to call that URL using https, second you need to pass the GMaps API key to it (or gmaps will reply back with an error saying _"Keyless access to Google Maps Platform is deprecated. Please use an API key with all your API calls to avoid service interruption. For further details please refer to http://g.co/dev/maps-no-account"_)

**Comment by sandrotosi** [comment]
fixed with:

```
@@ -34,14 +35,14 @@ class GoogleMapPlotter(object):
         self._infowindows = []
 
     @classmethod
-    def from_geocode(cls, location_string, zoom=13):
-        lat, lng = cls.geocode(location_string)
-        return cls(lat, lng, zoom)
+    def from_geocode(cls, location_string, apikey, zoom=13):
+        lat, lng = cls.geocode(location_string, apikey)
+        return cls(lat, lng, zoom, apikey)
 
     @classmethod
-    def geocode(self, location_string):
+    def geocode(self, location_string, apikey):
         geocode = requests.get(
-            'http://maps.googleapis.com/maps/api/geocode/json?address="%s"' % location_string)
+            'https://maps.googleapis.com/maps/api/geocode/json?address="%s"&key=%s' % (location_string, apikey))
         geocode = json.loads(geocode.text)
         latlng_dict = geocode['results'][0]['geometry']['location']
         return latlng_dict['lat'], latlng_dict['lng']
```

ie you're forced to pass and API key (as gmaps forces you to)

**Comment by imjackicle** [comment]
> fixed with:
> 
> ```
> @@ -34,14 +35,14 @@ class GoogleMapPlotter(object):
>          self._infowindows = []
>  
>      @classmethod
> -    def from_geocode(cls, location_string, zoom=13):
> -        lat, lng = cls.geocode(location_string)
> -        return cls(lat, lng, zoom)
> +    def from_geocode(cls, location_string, apikey, zoom=13):
> +        lat, lng = cls.geocode(location_string, apikey)
> +        return cls(lat, lng, zoom, apikey)
>  
>      @classmethod
> -    def geocode(self, location_string):
> +    def geocode(self, location_string, apikey):
>          geocode = requests.get(
> -            'http://maps.googleapis.com/maps/api/geocode/json?address="%s"' % location_string)
> +            'https://maps.googleapis.com/maps/api/geocode/json?address="%s"&key=%s' % (location_string, apikey))
>          geocode = json.loads(geocode.text)
>          latlng_dict = geocode['results'][0]['geometry']['location']
>          return latlng_dict['lat'], latlng_dict['lng']
> ```
> ie you're forced to pass and API key (as gmaps forces you to)

It worked for me. Thank you !

**Comment by hlongmore** [comment]
Any chance this can get merged into the official release? Or do I need to fork yet another project to fix bugs that have had a solution sitting around in some other fork for a while?

**Comment by frslm** [comment]
Fixed in #111.

**Comment by mriffey** [comment]
> Fixed in #111.

Thank you.
